hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MethodProbe.java

```diff
@@ -16,74 +16,154 @@
+ * A {@link MethodProbe} is a {@link ProbeFunction} that invokes a method that is annotated with {@link Probe}.
+ * <p>
+ * Internally accesses the method using reflection via one of these mechanisms (as a pose to a simple {@link Method#invoke} for
+ * <a href="https://github.com/hazelcast/hazelcast/pull/25279">performance reasons</a>):
+ * <ul>
+ * <li>{@link #methodHandle}
+ * <li>{@link #staticAccessor}
+ * <li>{@link #nonStaticAccessor}
+ * </ul>
  */
 abstract class MethodProbe implements ProbeFunction {
-
-    private static final Object[] EMPTY_ARGS = new Object[0];
-
-    final Method method;
+    private static final Lookup LOOKUP = MethodHandles.lookup();
+
+    /**
+     * {@link MethodHandle} used to access primitives to avoid boxing, specifically to reduce redundant object creation.
+     * <p>
+     * E.G. for a method that returns {@link long}, when typically invoked via reflection, an {@link Object} would instead be
+     * returned. The {@link long} would be boxed to {@link Long}, and then immediately unboxed again back to a {@link long} for
+     * returning from {@link LongProbeFunction#get(S)}.
+     * <p>
+     * Faster than basic reflection, but slower than a {@link LambdaMetafactory} generated accessor.
+     */
+    final MethodHandle methodHandle;
+
+    /**
+     * A {@link Supplier} used to access static methods returning objects (reference types).
+     * <p>
+     * Generated via a {@link LambdaMetafactory}, bound to a {@link MethodHandle} derived from {@link #methodHandle}.
+     * <p>
+     * Faster than basic reflection and {@link MethodHandle#invokeExact()}, but only applicable for reference types.
+     *
+     * @see <a href= "https://www.optaplanner.org/blog/2018/01/09/JavaReflectionButMuchFaster.html"> Further reading</a>
+     */
+    final Supplier<?> staticAccessor;
+    /**
+     * A {@link Function} used to access instance methods returning objects (reference types)
+     * <p>
+     * {@link Function#apply(T)}'s parameter is the instance to retrieve the value from.
+     *
+     * @see #staticAccessor
+     */
+    final Function<Object, ?> nonStaticAccessor;
+
+    final boolean isMethodStatic;
     final CachedProbe probe;
-    final int type;
+    final ProbeType type;
     final SourceMetadata sourceMetadata;
     final String probeName;
 
-    MethodProbe(Method method, Probe probe, int type, SourceMetadata sourceMetadata) {
-        this.method = method;
-        this.probe = new CachedProbe(probe);
-        this.type = type;
-        this.sourceMetadata = sourceMetadata;
-        this.probeName = probe.name();
-        assert probeName != null;
-        assert probeName.length() > 0;
-        method.setAccessible(true);
+    @SuppressWarnings("checkstyle:executablestatementcount")
+    MethodProbe(final Method method, final Probe probe, final ProbeType type, final SourceMetadata sourceMetadata) {
```

Comment:
Why are all these finals added? This isn't a style we generally use.

Reply:
> Why are all these finals added? This isn't a style we generally use.

I didn't see any style guides on use (or non-use) of `final`. But I can remove if an issue...

Reply:
We should use it consistently through the codebase unless there is a clear reason not to do it.

Reply:
> We should use it consistently through the codebase unless there is a clear reason not to do it.

Sorry I'm confused - are you saying we **should** use it, or it should be removed?

Reply:
IMHO it should be removed so that it is consistent with the rest of the code. Personally, I find the longer code not worth the additional value.

Reply:
63ce5646f1565d3c99a16efcd7aaa5f76d4d6f6b